### PR TITLE
refactor(server): library syncing

### PR DESCRIPTION
**Pull request by etnoy**
Merge all scan options into one
image

Library assets that are offline are moved to trash, with a little warning message attached to each asset
image

If the asset file is restored, the asset will be restored on next scan. If not, it'll be removed by the trash after the usual 30 days.

We remove isOffline and add a trashReason field to the asset entity

### CODE REVIEW

**Comment by jrasm91** [comment]
I think it is time to consolidate the library scanning jobs. I don't think we need 4 separate jobs in a context menu like this.

**Comment by jrasm91** [comment]
Can we have a single scan button/menu item that's opens a modal to submit a new scan job. It can use a drop-down to pick between a few options:
- new/added
- updated
- removed
- all

I think we can have a single library scanning job that takes in a body of options, that can use the selected type. Something like `{ type: 'new' }

**Comment by etnoy** [comment]
> Can we have a single scan button/menu item that's opens a modal to submit a new scan job. It can use a drop-down to pick between a few options:
> 
> * new/added
> * updated
> * removed
> * all
> 
> I think we can have a single library scanning job that takes in a body of options, that can use the selected type. Something like `{ type: 'new' }`

The pr description is already outdated, I've cleaned it up somewhat. But I agree 100% and will try to incorporate this as well

**Comment by mertalev** [comment]
I'd rather keep `isOffline` and use it for all assets instead of just library assets. It can be useful for the repair page, which is basically inaccessible for a lot of instances right now due to the amount of IO it does on each load.

**Comment by etnoy** [comment]
> I'd rather keep `isOffline` and use it for all assets instead of just library assets. It can be useful for the repair page, which is basically inaccessible for a lot of instances right now due to the amount of IO it does on each load.
> 
> Edit: also, isn't it a little dangerous if someone happens to make a mistake when changing their mounts or something goes offline and a scan wipes out all their asset info?

Regarding the danger part, I've been thinking about this for a while. With isOffline, we aren't really helping the user at the moment. Only when viewing individual assets could you potentially spot that things are offline due to the red box on the right hand side, and this is only on web, not mobile. If someone (including myself) removes assets, they'll just run scan immediately followed by remove offline without any hesitation between the two steps.

This change ensures that the removal of deleted assets only happens manually, it's never automatically scheduled in any way. I'll add a simple check to each import path to make sure it's still there. This way we can exclude the simple cases of a network share being unavailable and make sure it doesn't cause the whole library to be deleted.

**Comment by bo0tzz** [comment]
cc @zackpollard I think you were working on the repair page stuff right?

**Comment by zackpollard** [comment]
> cc @zackpollard I think you were working on the repair page stuff right?

Yea, I put it on hold when I saw this rework was happening, I will continue after this gets merged :smile:

**Comment by etnoy** [comment]
> > cc @zackpollard I think you were working on the repair page stuff right?
> 
> Yea, I put it on hold when I saw this rework was happening, I will continue after this gets merged 😄

You guys are talking about repair and I have no idea why. Are we using isOffline outside of external libraries? There's nothing in the code to suggest that

**Comment by zackpollard** [comment]
> > > cc @zackpollard I think you were working on the repair page stuff right?
> > 
> > 
> > Yea, I put it on hold when I saw this rework was happening, I will continue after this gets merged 😄
> 
> You guys are talking about repair and I have no idea why. Are we using isOffline outside of external libraries? There's nothing in the code to suggest that

The plan was to change the repair page to run in the background, and given we already have a lot of logic for scanning "external" libraries, it'd make sense to extend this to all libraries and use that logic for the repair page. Otherwise we are maintaining two implementations of scanning for missing files.

File: server/src/dtos/library.dto.ts
Comment (jrasm91) [inline server/src/dtos/library.dto.ts:100]: All of these options seems to be mutually exclusive, especially given how they are used in the web. Can we migrate this to an enum instead?

```ts
export enum LibraryAssetScan {
  ADDED='added', // files on disk without assets are added
  REMOVED='removed', // assets without files on disk are removed
  UPDATED='updated', // assets with files that have updated modification times are refreshed
}

Technically refreshing all assets isn't a scan, and I think it could be a separate endpoint. The implementation just needs to queue all assets for the library. You could argue that `libraryId` is an optional property to the "startJob" + metadata extraction job.

If we want to leave it in as another enum under `REFRESH='refresh'` that would be fine with me. 
```

**Comment by etnoy** [comment]
This PR has taken another turn, now it moves offline files to trash instead of deleting them. Four scan options have been merged into one

File: server/src/interfaces/asset.interface.ts
Comment (danieldietzler) [inline server/src/interfaces/asset.interface.ts:150]: Did you sort these alphabetically? 😅
It's a mess in the diff lol

File: server/src/interfaces/search.interface.ts
Comment (danieldietzler) [inline server/src/interfaces/search.interface.ts:66]: Why is it nullable here? Where's the difference between undefined and null for `trashReason` here?

File: server/src/repositories/asset.repository.ts
Comment (danieldietzler) [inline server/src/repositories/asset.repository.ts:297]: Why not just work with `undefined` in this case as well? IMO `this.updateAll(ids, {})` is a bit simpler

File: server/src/services/asset.service.ts
Comment (danieldietzler) [inline server/src/services/asset.service.ts:291]: I'd prefer just destructuring `trashReason` in line 278 as well.

File: server/src/services/library.service.spec.ts
Comment (danieldietzler) [inline server/src/services/library.service.spec.ts:229]: Can't you just `.mockImplementation(() => {})`?

File: server/src/services/library.service.spec.ts
Comment (danieldietzler) [inline server/src/services/library.service.spec.ts:532]:
```suggestion
    it('should not refresh a file which mtime matches existing asset', async () => {
```

File: server/src/services/library.service.spec.ts
Comment (danieldietzler) [inline server/src/services/library.service.spec.ts:553]:
```suggestion
    it('should refresh a file which mtime differs from existing asset', async () => {
```

File: server/src/services/library.service.ts
Comment (danieldietzler) [inline server/src/services/library.service.ts:377]: I still find this quite hard to read/complex overall (lots of nesting/conditions).

File: server/src/services/library.service.ts
Comment (danieldietzler) [inline server/src/services/library.service.ts:384]: You don't need the else here.

File: server/src/services/library.service.ts
Comment (danieldietzler) [inline server/src/services/library.service.ts:390]: You don't need the else here.

File: server/src/services/library.service.ts
Comment (danieldietzler) [inline server/src/services/library.service.ts:395]: You don't need this else.

File: server/src/services/trash.service.ts
Comment (danieldietzler) [inline server/src/services/trash.service.ts:27]:
```suggestion
    const deletedAssets = assets.filter(({ trashReason }) => trashReason === AssetTrashReason.DELETED);
```

File: server/src/services/trash.service.ts
Comment (danieldietzler) [inline server/src/services/trash.service.ts:43]:
```suggestion
      const ids = assets.filter(({ trashReason }) => trashReason === AssetTrashReason.DELETED).map((a) => a.id);
```

File: web/src/lib/components/asset-viewer/asset-viewer-nav-bar.svelte
Comment (danieldietzler) [inline web/src/lib/components/asset-viewer/asset-viewer-nav-bar.svelte:140]:
```suggestion
        {#if asset.trashReason === TrashReason.Deleted}
```

File: web/src/lib/components/asset-viewer/detail-panel.svelte
Comment (danieldietzler) [inline web/src/lib/components/asset-viewer/detail-panel.svelte:156]: I think with regards to RTL languages this entire string needs to be localized (so passing the path as a variable), as the structure `Path: /path/to/asset` may not always work

**Review by danieldietzler — COMMENTED** [review]
Regarding the translation files: I would just update the English file. Then, weblate will delete the keys in all the other files and add new ones, yes, but that may actually be what we want as translations may be inaccurate now.

File: server/src/services/library.service.ts
Comment (etnoy) [reply server/src/services/library.service.ts:395]: It's in there for the log statement

File: server/src/services/trash.service.ts
Comment (etnoy) [reply server/src/services/trash.service.ts:43]: Thanks. Do you know of a way to do the filter already in the db call? I couldn't find a good way

File: server/src/services/library.service.ts
Comment (danieldietzler) [reply server/src/services/library.service.ts:395]: You always return before, that's why you don't need the else

File: server/src/services/trash.service.ts
Comment (danieldietzler) [reply server/src/services/trash.service.ts:43]: cc @mertalev

File: server/src/controllers/library.controller.ts
Comment (etnoy) [reply server/src/controllers/library.controller.ts:69]: It's an admin-only route so it should be fine :)

File: server/src/interfaces/search.interface.ts
Comment (etnoy) [reply server/src/interfaces/search.interface.ts:66]: Fixed!

File: server/src/interfaces/asset.interface.ts
Comment (etnoy) [reply server/src/interfaces/asset.interface.ts:150]: Maybe I should move the sorting of repos to another PR. I can't stand them unsorted :)

File: server/src/services/asset.service.ts
Comment (etnoy) [reply server/src/services/asset.service.ts:291]: Done

File: server/src/services/library.service.spec.ts
Comment (etnoy) [reply server/src/services/library.service.spec.ts:229]: No, it didn't work out like that

File: server/src/services/library.service.spec.ts
Comment (etnoy) [reply server/src/services/library.service.spec.ts:532]: Nope, my version is correct :)

File: server/src/services/library.service.ts
Comment (etnoy) [reply server/src/services/library.service.ts:395]: Got it!

File: server/src/services/library.service.ts
Comment (etnoy) [reply server/src/services/library.service.ts:390]: Done

File: server/src/repositories/asset.repository.ts
Comment (danieldietzler) [reply server/src/repositories/asset.repository.ts]: Nvm, has been discussed on Discord.

File: server/src/controllers/library.controller.ts
Comment (bo0tzz) [reply server/src/controllers/library.controller.ts:69]: We probably still want to have permission checks, especially if we're moving towards applying libraries more broadly, right?

Comment (etnoy) [reply server/src/controllers/library.controller.ts:69]: Good point, I'm adding the LIBRARY_UPDATE permission to the endpoint

File: web/src/lib/components/asset-viewer/detail-panel.svelte
Comment (etnoy) [reply web/src/lib/components/asset-viewer/detail-panel.svelte:156]: Fixed now

File: server/src/services/trash.service.ts
Comment (zackpollard) [reply server/src/services/trash.service.ts:43]: Should just be able to add `trashReason: AssetTrashReason.DELETED,` below line 38, right?

File: server/src/services/library.service.ts
Comment (zackpollard) [inline server/src/services/library.service.ts:524]: Not sure how I feel about this specifically, it leaves a bit of an open edge-case that won't be picked up until the next scan. Not sure how we'd resolve that really though while still using the trashReason stuff.

**Review by zackpollard — DISMISSED** [review]
Some notes, but overall this looks like a great change and vastly simplifies the external libraries scanning code. Once comments are resolved I am happy for this to be merged in :smile:

File: server/src/entities/asset.entity.ts
Comment (mertalev) [reply server/src/entities/asset.entity.ts:176]: I think it makes sense for it to be nullable, personally. This is something that's only really known when it's being trashed, so it should be set at that time.

Comment (zackpollard) [reply server/src/entities/asset.entity.ts:176]: Right, so it should be set to deleted. Everything that exists right now is deleted. What's the reason for not setting it?

Comment (etnoy) [reply server/src/entities/asset.entity.ts:176]: You bring up a good point @zackpollard. Any existing assets that are in the trash must be migrated to have trashReason set to deleted.

I also agree with @mertalev that trashReason should be nullable since the field will be null until it's trashed.

Therefore, I need to write a migration to set the field for all deleted assets.

Comment (mertalev) [reply server/src/entities/asset.entity.ts:176]: Are you talking about the migration? The value it should be set to depends on whether `deletedAt` or `isOffline` is true, so there should be a query that sets it with `case when "deletedAt" is not null then 'deleted' when "isOffline" then 'offline' end`. In the case where `isOffline` is true and `deletedAt` is null, it should also set `deletedAt`.

**Comment by etnoy** [comment]
> Some notes, but overall this looks like a great change and vastly simplifies the external libraries scanning code. Once comments are resolved I am happy for this to be merged in 😄

Thanks! Happy that you like the concepts in this PR. I'll do some modifications based on your suggestions.

File: server/src/entities/asset.entity.ts
Comment (zackpollard) [reply server/src/entities/asset.entity.ts:176]: Honestly completely overlooked the case of it not being trashed 🤣 But yes we can do that for the existing trashed assets was more my point :P

File: server/src/services/trash.service.ts
Comment (etnoy) [reply server/src/services/trash.service.ts]: Done!

File: mobile/lib/extensions/collection_extensions.dart
Comment (etnoy) [reply mobile/lib/extensions/collection_extensions.dart:75]: Yes, but I have no idea if I can remove the nonOfflineOnly function

File: server/src/services/library.service.ts
Comment (etnoy) [reply server/src/services/library.service.ts:373]: You're right, the log messages I added make the comments redundant

File: server/src/services/library.service.ts
Comment (etnoy) [reply server/src/services/library.service.ts:524]: The whole reason for the trashReason complexity is to handle this edge case. What edge case is left open?

File: server/src/services/library.service.ts
Comment (etnoy) [reply server/src/services/library.service.ts]: Done

File: server/src/entities/asset.entity.ts
Comment (jrasm91) [reply server/src/entities/asset.entity.ts:176]: The question seems to be "do we want `null` to imply `deleted`". I am fine making it non-nullable, always requiring a value, and setting it to `deleted` when a request comes in through the api.

Comment (mertalev) [reply server/src/entities/asset.entity.ts:176]: With kysely, null fields can be omitted from DB responses entirely, so it's generally better to set something null than to have a default that has the same meaning.

Comment (jrasm91) [reply server/src/entities/asset.entity.ts:176]: Very free queries return deleted data in the first place, so I don't think that's a very strong argument.

Comment (mertalev) [reply server/src/entities/asset.entity.ts:176]: But if `trashReason` were always set to a value, it'd always be part of the response even if the asset isn't trashed. I guess you could omit it from the select instead, but that's not as clean.

Comment (jrasm91) [reply server/src/entities/asset.entity.ts:176]: Oh right. I think I meant to recommend it always have a value _when an asset has been deleted_. So null when not deleted, deleted when trashed and offline when it goes offline. I guess that means nullable, but not "null implies deleted".

File: server/src/services/library.service.ts
Comment (zackpollard) [reply server/src/services/library.service.ts:519]: If an asset is trashed and goes offline what happens? We don't change the reason right? So if the user then untrashes the asset it could be in their timeline until the next scan but be broken. An edge case though and not something we can easily mitigate.

Comment (jrasm91) [reply server/src/services/library.service.ts:519]: Hah, I guess we could make another enum "trashed _and_ offline"

File: server/src/entities/asset.entity.ts
Comment (mertalev) [reply server/src/entities/asset.entity.ts:176]: Ah, that makes sense. I also agree with that.

Comment (mertalev) [reply server/src/entities/asset.entity.ts:176]: Maybe there should be a check constraint on this like `("deletedAt" is null) = ("trashReason" is null)`, so we have an invariant that either both are null or neither are.

Comment (etnoy) [reply server/src/entities/asset.entity.ts:176]: Just to confirm: I currently implemented this by having trashReason be null for all non-trashed item. When an item is trashed by the api, it's set to deleted, and when the file is offline it's set to offline. This is the way we want, right?

Comment (zackpollard) [reply server/src/entities/asset.entity.ts:176]: Yes that sounds right to me 🙂

File: server/src/interfaces/job.interface.ts
Comment (danieldietzler) [inline server/src/interfaces/job.interface.ts:83]: Nit: can we use the enum names as the values instead of using new words? To me this is somewhat confusing.

File: server/src/services/library.service.ts
Comment (jrasm91) [inline server/src/services/library.service.ts:246]: Maybe we should rename this to syncFiles? Also maybe `syncFiles(assetPaths: string, { libraryId, ownerId }: { libraryId: string, ownerId: string })` would look better

File: server/src/services/library.service.ts
Comment (jrasm91) [inline server/src/services/library.service.ts:556]: Maybe just `count`.

**Review by jrasm91 — APPROVED** [review]
Thanks for working on this @etnoy and @zackpollard!

File: server/src/services/library.service.ts
Comment (danieldietzler) [inline server/src/services/library.service.ts:385]: Shouldn't we check if the library actually exists? I guess it's already covered wherever we queue the sync, but it still feels odd not sanity-checking that here

File: server/src/services/library.service.ts
Comment (danieldietzler) [inline server/src/services/library.service.ts:565]: I'd prefer being explicit here, so `if (count > 0)`

**Review by danieldietzler — APPROVED** [review]

**Pull request merged**
